### ISSUE TITLE
Huge memory consumption with quantities higher than 30.000, running in Out of Memory exception

### ISSUE THREAD

**Issue by itinance** [Sylius/Sylius#10741]
**Sylius version affected**: 1.5

**Description**  

There is a weird thing with the quantity of an OrderItem depending on the quantity itself consuming too much memory depending on the actual value. After I patched the system-wide hard coded limitation of 9990 for any quantity (as reported in #10740) it becomes really strange:

Ordering 15.000 (15k) items of an article is working well, 25k also, but any other number higher than 30.000 will result in Out of Memory exception (my Limit is at 250 MB). Although I can increase `memory_limit`, what is going on there? It feels like one would create an array consisting of `$quantity` count of items with much memory consumption each, which will cause this trouble depending on the actual value in the quantity.

The threshold with my personal `memory_limit` is between 30k and 50k passed as quantity. Sometimes 45k went through successfully, but mostly it crashes with `Out of Memory` with any number higher than 30k, but never below 30k. 

**Steps to reproduce**  

1. Patch the [service definition](https://github.com/Sylius/Sylius/blob/23aeeb5e6b563518192c79a2bc768e254d9ea9ad/src/Sylius/Bundle/CoreBundle/Resources/config/services.xml#L158) or the code behind in [LimitingOrderItemQuantityModifier.php](https://github.com/Sylius/Sylius/blob/583f53aa5a19d5203db8e9ac0f8af9370b2fcb57/src/Sylius/Component/Core/Cart/Modifier/LimitingOrderItemQuantityModifier.php#L38)  in order to allow quantity values higher then 30000 

2. submit orders with quantity of > 30.000, e.g. 40k, 50k, 100k while `memory_limit` was set to 250 MB.

**Comment by vvasiloi** [comment]
https://github.com/Sylius/Sylius/issues/10155

**Comment by lchrusciel** [comment]
Hey @itinance,

your memory consumption problems may be related to the fact that the quantity of items depends on item units, which means that ordering 30k of something will create 30k objects in the database (OrderItemUnits). 
Depending on how your checkout process looks like, you may change these units on every cart operation or every step of the checkout. Therefore memory consumption may be really high.

There is no solution I can fastly come up with, as this is a root behavior of Sylius.

**Comment by CoderMaggie** [comment]
I totally get the idea of testing the limits of the system...
but could you provide me REAL use case where a customer buys 30k of units?

Maybe your assumptions are wrong, because I can't imagine a shop with possible scenarios of both buying a single item and 30k items. One of these items is sold in wrong quantities probably ;)

**Comment by CoderMaggie** [comment]
Correct me if I am thinking wrongly, but maybe it would be possible to introduce variants for batch quantities instead? 

variant A: 0.0001
variant B: 1000
variant C: 1 billion
and similar?

**Comment by itinance** [comment]
That sounds like a good idea. I will take it into further considerations! Many thanks.

**Comment by itinance** [comment]
Okay, after playing a lot with this issue I came across an ugly hack as possible solution for this.

Before this hack, I tried several other approaches like provide an own custom `QuantityModifier` in conjunction with own `OrderItemUnit`-entity, that was derived from the Original `OrderItemUnit` entity but handling the `total`-amount itself without creating thousands of `OrderItemUnit`s. 
Nothing has really worked properly.

At the end I ran into this issue, where my custom `OrderItemUnit` was not compatible for OrderItem's `unit`-ArrayCollection although it was derived from the Base class out of sylius:

[Expected value of type "Sylius\Component\Core\Model\OrderItemUnit" for association field "AppBundle\Entity\OrderItem#$units", got "AppBundle\Entity\TokenOrderItemUnit" instead.
](url)

Since:
- we don't need any partial adjustments on our sold items
- we sell huge numbers in quantity (mostly far more than 9999, see #10740)

we made a cut here with the following "hack":

When processing the order, we first always create an OrderItem with Quantity of 1.

After all, at the end of this process, we fix database tables with the actual values for `total` and `quantity` (and yes, we are aware what we're doing here and know that this is not the best practice ^^):

		/** @var OrderRepositoryInterface $orderRepository */
		$orderRepository = $this->container->get('sylius.repository.order');
		$orderRepository->add($order);
		$this->container->get('sylius.order_processing.order_processor')->process($order);

		$this->entityManager->flush();
		$conn = $this->entityManager->getConnection();

		$stmt = $conn->prepare('UPDATE sylius_order SET items_total=:total, total=:total WHERE id=:id');
		$stmt->bindValue('total', $amount * 100);
		$stmt->bindValue('id', $order->getId());
		$stmt->execute();

		$stmt = $conn->prepare('UPDATE sylius_order_item SET quantity=:total, units_total=:units_total, total=:units_total WHERE order_id=:id');
		$stmt->bindValue('total', $amount);
		$stmt->bindValue('units_total', $amount * 100);
		$stmt->bindValue('id', $order->getId());
		$stmt->execute();


This way we bypass both issues #10741 and #10740 and according our tests, it seems to work fine.
